### PR TITLE
Fixes #4071 - Service Port validations are not consistently enforced

### LINKED ISSUE #4071: Service Port validations are not consistently enforced
## Creating multiple apps in the same group with a single request

`POST /v2/groups/` the following:

<details>
```json
{
  "id": "/abc",
  "apps": [
    {
      "id": "app1",
      "cmd": "sleep 500",
      "container": {
        "type": "DOCKER",
        "docker": {
          "image": "busybox",
          "network": "BRIDGE",
          "portMappings": [
            {
              "containerPort": 5000,
              "hostPort": 0,
              "servicePort": 16554
            }
          ]
        }
      },
      "mem": 32,
      "cpus": 0.5,
      "instances": 0
    },
    {
      "id": "app2",
      "cmd": "sleep 500",
      "container": {
        "type": "DOCKER",
        "docker": {
          "image": "busybox",
          "network": "BRIDGE",
          "portMappings": [
            {
              "containerPort": 5000,
              "hostPort": 0,
              "servicePort": 16554
            }
          ]
        }
      },
      "mem": 32,
      "cpus": 0.5,
      "instances": 0
    }
  ]
}
```
</details>

You will get:

```json
{
    "details": [
        {
            "errors": [
                "Requested service port 16554 conflicts with a service port in app /abc/app2",
                "Requested service port 16554 conflicts with a service port in app /abc/app1"
            ],
            "path": "/groups(0)/value"
        }
    ],
    "message": "Object is not valid"
}
```
This is the desired behaviour.

## Create multiple apps in the same group using the `POST /v2/apps` endpoint
`POST /v2/apps` with:

<details>
```json
[
  {
    "id": "/abc/app1",
    "cmd": "sleep 500",
    "container": {
      "type": "DOCKER",
      "docker": {
        "image": "busybox",
        "network": "BRIDGE",
        "portMappings": [
          {
            "containerPort": 5000,
            "hostPort": 0,
            "servicePort": 16554
          }
        ]
      }
    },
    "mem": 32,
    "cpus": 0.5,
    "instances": 0
  }
```
</details>

And then `POST /v2/apps`:

<details>
```json
  {
    "id": "/abc/app2",
    "cmd": "sleep 500",
    "container": {
      "type": "DOCKER",
      "docker": {
        "image": "busybox",
        "network": "BRIDGE",
        "portMappings": [
          {
            "containerPort": 5000,
            "hostPort": 0,
            "servicePort": 16554
          }
        ]
      }
    },
    "mem": 32,
    "cpus": 0.5,
    "instances": 0
  }
]
```
</details>

We get the same validation error, which is good.

## Creating apps in different groups using the `PUT /v2/groups` endpoint

PUT `/v2/groups`:

<details>
```json
{
  "id": "/",
  "groups": [
    {
      "id": "/abc",
      "apps": [
        {
          "id": "app1",
          "cmd": "sleep 500",
          "container": {
            "type": "DOCKER",
            "docker": {
              "image": "busybox",
              "network": "BRIDGE",
              "portMappings": [
                {
                  "containerPort": 5000,
                  "hostPort": 0,
                  "servicePort": 16554
                }
              ]
            }
          },
          "mem": 32,
          "cpus": 0.5,
          "instances": 0
        }
      ]
    },
    {
      "id": "/def",
      "apps": [
        {
          "id": "app2",
          "cmd": "sleep 500",
          "container": {
            "type": "DOCKER",
            "docker": {
              "image": "busybox",
              "network": "BRIDGE",
              "portMappings": [
                {
                  "containerPort": 5000,
                  "hostPort": 0,
                  "servicePort": 16554
                }
              ]
            }
          },
          "mem": 32,
          "cpus": 0.5,
          "instances": 0
        }
      ]
    }
  ]
}
```
</details>

Is allowed, but it should **NOT** be.

## Creating apps in different groups using the `PUT /v2/apps` endpoint

 `PUT /v2/apps`:

<details>
```json
[
  {
    "id": "/abc/app1",
    "cmd": "sleep 500",
    "container": {
      "type": "DOCKER",
      "docker": {
        "image": "busybox",
        "network": "BRIDGE",
        "portMappings": [
          {
            "containerPort": 5000,
            "hostPort": 0,
            "servicePort": 16554
          }
        ]
      }
    },
    "mem": 32,
    "cpus": 0.5,
    "instances": 0
  },
  {
    "id": "/def/app2",
    "cmd": "sleep 500",
    "container": {
      "type": "DOCKER",
      "docker": {
        "image": "busybox",
        "network": "BRIDGE",
        "portMappings": [
          {
            "containerPort": 5000,
            "hostPort": 0,
            "servicePort": 16554
          }
        ]
      }
    },
    "mem": 32,
    "cpus": 0.5,
    "instances": 0
  }
]
```
</details>

Is allowed, but it should **NOT** be.

## Creating apps in different groups using two POST `/v2/groups` calls

<details>
`POST /v2/groups`
```json
{
  "id": "/abc",
  "apps": [
    {
      "id": "app1",
      "cmd": "sleep 500",
      "container": {
        "type": "DOCKER",
        "docker": {
          "image": "busybox",
          "network": "BRIDGE",
          "portMappings": [
            {
              "containerPort": 5000,
              "hostPort": 0,
              "servicePort": 16554
            }
          ]
        }
      },
      "mem": 32,
      "cpus": 0.5,
      "instances": 0
    }
  ]
}
```

And then `POST /v2/groups`:

```json
{
  "id": "/def",
  "apps": [
    {
      "id": "app2",
      "cmd": "sleep 500",
      "container": {
        "type": "DOCKER",
        "docker": {
          "image": "busybox",
          "network": "BRIDGE",
          "portMappings": [
            {
              "containerPort": 5000,
              "hostPort": 0,
              "servicePort": 16554
            }
          ]
        }
      },
      "mem": 32,
      "cpus": 0.5,
      "instances": 0
    }
  ]
}
```
</details>

Is allowed, but it should **NOT** be.

### PR DESCRIPTION
**Pull request by gkleiman**

### CODE REVIEW

File: src/test/scala/mesosphere/marathon/integration/NetworkPartitionIntegrationTest.scala [hunk 1]
```diff
@@ -2,7 +2,7 @@ package mesosphere.marathon.integration
 
 import mesosphere.marathon.integration.facades.ITEnrichedTask
 import mesosphere.marathon.integration.setup._
-import org.scalatest.{BeforeAndAfter, GivenWhenThen, Matchers}
+import org.scalatest.{ BeforeAndAfter, GivenWhenThen, Matchers }
```
Comment (jdef) [inline src/test/scala/mesosphere/marathon/integration/NetworkPartitionIntegrationTest.scala]: what does this have to do with the intent of this PR?

Comment (gkleiman) [inline src/test/scala/mesosphere/marathon/integration/NetworkPartitionIntegrationTest.scala]: This was automatically done by sbt when running our tests, because the current state of the repo doesn't follow the style rules.

https://github.com/mesosphere/marathon/pull/4068 has the same change. I'll merge that PR.

**Comment by gkleiman** [comment]
This should also fix #3999

File: src/main/scala/mesosphere/marathon/state/Group.scala [hunk 2]
```diff
@@ -254,19 +254,19 @@ object Group {
...
-              existingServicePort <- existingApp.container.flatMap(_.portMappings).toList.flatten.map(_.servicePort)
-              if existingServicePort != 0 // ignore zero ports, which will be chosen at random
-              if servicePorts contains existingServicePort
-            } yield RuleViolation(
-              app.id,
-              s"Requested service port $existingServicePort conflicts with a service port in app ${existingApp.id}",
-              None)
-          }
+        val groupViolations = group.transitiveApps.flatMap { app =>
+          val servicePorts = app.servicePorts.toSet
```
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/state/Group.scala]: The old code used `app.container.flatMap(_.servicePorts)` while container.servicePorts returns `container.flatMap(_.servicePorts).getOrElse(portNumbers)`. I am not too familiar with the underlying implementation – is does that make sense?

File: src/main/scala/mesosphere/marathon/state/Group.scala [hunk 3]
```diff
@@ -254,19 +254,19 @@ object Group {
...
+
+          val ruleViolations = for {
+            existingApp <- group.transitiveApps
```
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/state/Group.scala]: You're fetching group.transitiveApps here the second time (see line 257).

File: src/main/scala/mesosphere/marathon/state/Group.scala (hunk 2 above)
Comment (gkleiman) [inline src/main/scala/mesosphere/marathon/state/Group.scala]: The old code was actually buggy and wouldn't validate service ports defined through `portDefinitions`. This change fixes that.

Comment (jdef) [inline src/main/scala/mesosphere/marathon/state/Group.scala]: nice catch @gkleiman

**Comment by aquamatthias** [comment]
LGTM. Should be cherry picked into 1.2
